video/out/w32_common.c

```diff
@@ -974,7 +991,19 @@ static void reinit_window_state(struct vo_w32_state *w32)
 
     // Assume that the window has already been fit on screen before switching fs
     if (!toggle_fs || w32->fit_on_screen) {
+        // FIXME: fit_window_on_screen ensures the content (and decorations if
+        // --fit-border is used) is contained entirely inside the screen.
+        // Therefore, it's important that it's not called when the user expects
+        // to restore a size/pos which was set manually, even if it's bigger
+        // and/or partially outside the screen, which is applicable at least
+        // when restoring from fullscreen/maximized/minimized/sleep/etc.
+        // This does seem to work correctly - empirically, however, the
+        // conditions which prevent calling it in these cases are delicate and
+        // hard to follow, with various state flags.
+        // We need to make it more obvious/documented how this call is avoided
+        // on such cases (i.e. how the next line is not reached).
```

Comment:
So, I assume from this comment, do you need to adjust the if statement above to skip calling the `fit_window_on_screen` in some more cases? I don't quite understand the problem here to be honest.

Reply:
`fit_window_on_screen` should not be called, unless there was a change in fullscreen state, or it was specifically asked to refit the window rect by setting the `w32->fit_on_screen`.

The `w32->fit_on_screen` is set to true when there's a change made to the window rect, so it needs to be refit on screen on next update.

Reply:
The scenario which we don't want to reach is that the user manually moves/resizes the window to a state which fit would change it if called.

For instance, the user moves the window such that part of it is off-screen. Then they enter fullscreen or maximize or minimize, and then restore the window.

We want to restore it to the same position - partially off screen, but if the fit function would be called, then it will re-position the window inside the screen - which is not what the user expects.

In my attempts to trigger this case I couldn't make it happen, but I also couldn't guarantee that it will never happen, because the states at the code were too complex for me to follow.

So I'm asking whether it's safe to assume that we will not move the window into the screen while the user expects it to be restored to its original position.

Now is different than before because we now move the window if it's partially off screen, while previously fit only had an effect when the window size was too big.

Reply:
You could try running mpv with `geometry` option set, and see what will happen when you leave fullscreen.

Or you can try to trigger the conditions when `VOCTRL_SET_UNFS_WINDOW_SIZE` request is sent.

As far as I can see there are only 4 cases when `reinit_window_state` is called:
1) When restoring the maximized window in fullscreen state. (the `fit_window_on_screen` do nothing for maximized window anyway)
2) When there's a fullscreen state change detected (`w32->fit_on_screen` is not set in this case anyway).
3) In `gui_thread_reconfig` which is called on vo->reconfigure I suppose.
4) When the window size is changed via `VOCTRL_SET_UNFS_WINDOW_SIZE` request. This is the only case among 4 when `w32->fit_on_screen` flag could make an effect.

Reply:
> You could try running mpv with geometry ...
> Or you can try to trigger ...
> As far as I can see there are only 4 cases when reinit_window_state is called ...

Yeah, I know where it's called from, and I tried to trigger it in many ways, including multi-monitor setup in both horizontal and vertical configuration, and various other cases.

But ultimately that's anecdotal, because it's without deep understanding of how the states are updated, which WM_* handlers are called in which order, etc.

The point of the question was to understand whether you, as the original author, know, or have a good assessment of whether or not it could hit undesirable invocations.